pkg/jobrunaggregator/jobrunaggregatorlib/ci_data_client.go

```diff
@@ -0,0 +1,317 @@
+	"google.golang.org/api/iterator"
+)
+
+type CIDataClient interface {
+	ListAllJobs(ctx context.Context) ([]jobrunaggregatorapi.JobRow, error)
+
+	// GetJobRunForJobNameBeforeTime returns the jobRun closest to, but BEFORE, the time provided.
+	// This is useful for bounding a query of GCS buckets in a window.
+	// nil means that no jobRun was found before the specified time.
+	GetJobRunForJobNameBeforeTime(ctx context.Context, jobName string, targetTime time.Time) (*jobrunaggregatorapi.JobRunRow, error)
+
+	// GetJobRunForJobNameAfterTime returns the jobRun closest to, but AFTER, the time provided.
+	// This is useful for bounding a query of GCS buckets in a window.
+	// nil means that no jobRun as found after the specified time.
+	GetJobRunForJobNameAfterTime(ctx context.Context, jobName string, targetTime time.Time) (*jobrunaggregatorapi.JobRunRow, error)
+
+	GetLastJobRunWithTestRunDataForJobName(ctx context.Context, jobName string) (*jobrunaggregatorapi.JobRunRow, error)
+	GetLastJobRunWithDisruptionDataForJobName(ctx context.Context, jobName string) (*jobrunaggregatorapi.JobRunRow, error)
+	GetLastAggregationForJob(ctx context.Context, frequency, jobName string) (*jobrunaggregatorapi.AggregatedTestRunRow, error)
+	ListUnifiedTestRunsForJobAfterDay(ctx context.Context, jobName string, startDay time.Time) (*UnifiedTestRunRowIterator, error)
+
+	ListAggregatedTestRunsForJob(ctx context.Context, frequency, jobName string, startDay time.Time) ([]jobrunaggregatorapi.AggregatedTestRunRow, error)
+}
+
+type ciDataClient struct {
+	dataCoordinates BigQueryDataCoordinates
+	client          *bigquery.Client
+}
+
+func NewCIDataClient(dataCoordinates BigQueryDataCoordinates, client *bigquery.Client) CIDataClient {
+	return &ciDataClient{
+		dataCoordinates: dataCoordinates,
+		client:          client,
+	}
+}
+
+func (c *ciDataClient) ListAllJobs(ctx context.Context) ([]jobrunaggregatorapi.JobRow, error) {
+	queryString := c.dataCoordinates.SubstituteDataSetLocation(
+		`SELECT *  
+FROM DATA_SET_LOCATION.Jobs
+ORDER BY Jobs.JobName ASC
+`)
+
+	query := c.client.Query(queryString)
+	jobRows, err := query.Read(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query job table with %q: %w", queryString, err)
+	}
+	jobs := []jobrunaggregatorapi.JobRow{}
+	for {
+		job := &jobrunaggregatorapi.JobRow{}
+		err = jobRows.Next(job)
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		jobs = append(jobs, *job)
+	}
+
+	return jobs, nil
+}
+
+func (c *ciDataClient) GetLastJobRunWithTestRunDataForJobName(ctx context.Context, jobName string) (*jobrunaggregatorapi.JobRunRow, error) {
+	queryString := c.dataCoordinates.SubstituteDataSetLocation(
+		`
+SELECT distinct(JobRuns.Name), JobRuns.StartTime
+FROM DATA_SET_LOCATION.JobRuns 
+INNER JOIN DATA_SET_LOCATION.TestRuns on TestRuns.JobRunName = JobRuns.Name
+WHERE JobRuns.JobName = @JobName
+ORDER BY JobRuns.Name DESC
```

Comment:
Was the order by here meant to be on start time? The function name indicates we're looking for most recent, but we query for results with a certain name, and then sort on name.

Reply:
Still confused about this query, why are we querying on JobName and then sorting on it? With the func name I'd think this was meant to be start time.

Reply:
> Still confused about this query, why are we querying on JobName and then sorting on it? With the func name I'd think this was meant to be start time.

Weird but true. The name of the job run has a guaranteed order.  The startTime of the prowjob itself is not guaranteed.